File: src/main/java/fr/maxlego08/satisfactorydle/command/EntityCache.java
package fr.maxlego08.satisfactorydle.command;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import fr.maxlego08.satisfactorydle.SatisfactorydleAPI;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;

public class EntityCache {

    private final SatisfactorydleAPI api;
    private final Map<String, List<EntityEntry>> cache = new ConcurrentHashMap<>();
    public EntityCache(SatisfactorydleAPI api) {
        this.api = api;
    }

    public String cacheKey(String mode, String locale) {
        return mode + ":" + locale;
    }

    public List<EntityEntry> get(String key) {
        return cache.get(key);
    }

    public void store(String key, JsonArray entitiesArray) {
        List<EntityEntry> entries = new ArrayList<>();
        for (JsonElement element : entitiesArray) {
            JsonObject obj = element.getAsJsonObject();
            entries.add(new EntityEntry(obj.get("id").getAsInt(), obj.get("name").getAsString()));
        }
        cache.put(key, entries);
    }

    public void ensureCache(String mode, String locale) throws Exception {
        String key = cacheKey(mode, locale);
        if (!cache.containsKey(key)) {
            JsonObject daily = api.getDaily(mode, locale);
            store(key, daily.getAsJsonArray("entities"));
        }
    }

    public void ensureCacheAsync(String mode, String locale) {
        String key = cacheKey(mode, locale);
        if (!cache.containsKey(key)) {
            CompletableFuture.runAsync(() -> {
                try {
                    JsonObject daily = api.getDaily(mode, locale);
                    store(key, daily.getAsJsonArray("entities"));
                } catch (Exception ignored) {
                }
            });
        }
    }

    public record EntityEntry(int id, String name) {
    }
}

File: app/commands/entity_cache.py
from __future__ import annotations

from dataclasses import dataclass

from app.api.satisfactorydle_api import SatisfactorydleAPI


@dataclass(slots=True, frozen=True)
class EntityEntry:
    id: int
    name: str


class EntityCache:
    def __init__(self, api: SatisfactorydleAPI) -> None:
        self.api = api
        self._cache: dict[str, list[EntityEntry]] = {}

    @staticmethod
    def cache_key(mode: str, locale: str) -> str:
        return f"{mode}:{locale}"

    def get(self, key: str) -> list[EntityEntry] | None:
        return self._cache.get(key)

    def store(self, key: str, entities_array: list[dict]) -> None:
        entries = [EntityEntry(id=int(entry["id"]), name=str(entry["name"])) for entry in entities_array]
        self._cache[key] = entries

    async def ensure_cache(self, mode: str, locale: str) -> None:
        key = self.cache_key(mode, locale)
        if key not in self._cache:
            daily = await self.api.get_daily(mode, locale)
            self.store(key, daily.get("entities", []))

    async def ensure_cache_async(self, mode: str, locale: str) -> None:
        key = self.cache_key(mode, locale)
        if key in self._cache:
            return
        try:
            daily = await self.api.get_daily(mode, locale)
            self.store(key, daily.get("entities", []))
        except Exception:  # noqa: BLE001
            return
